### PR TITLE
llama + spec: MTP support

### PR DESCRIPTION
**Pull request by am17an**
## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure:

### CODE REVIEW

File: common/arg.cpp [hunk 1]
```diff
@@ -3562,12 +3562,15 @@ common_params_context common_params_parser_init(common_params & params, llama_ex
         }
     ).set_spec().set_examples({LLAMA_EXAMPLE_SPECULATIVE, LLAMA_EXAMPLE_SERVER, LLAMA_EXAMPLE_CLI}));
     add_opt(common_arg(
-        {"--spec-type"}, "[none|ngram-cache|ngram-simple|ngram-map-k|ngram-map-k4v|ngram-mod]",
-        string_format("type of speculative decoding to use when no draft model is provided (default: %s)\n",
+        {"--spec-type"}, "[none|mtp|ngram-cache|ngram-simple|ngram-map-k|ngram-map-k4v|ngram-mod]",
+        string_format("type of speculative decoding to use when no draft model is provided (default: %s). "
+                      "'mtp' loads a sibling ctx_mtp from the same .gguf with override_arch=qwen35_mtp (Qwen3.5/3.6).\n",
```
Comment (am17an) [inline common/arg.cpp]: don't need to add this comment because it is superflous

File: common/common.h [hunk 2]
```diff
@@ -159,6 +159,7 @@ enum common_speculative_type {
     COMMON_SPECULATIVE_TYPE_NONE,          // no speculative decoding
     COMMON_SPECULATIVE_TYPE_DRAFT,         // draft model
     COMMON_SPECULATIVE_TYPE_EAGLE3,        // eagle draft model
+    COMMON_SPECULATIVE_TYPE_MTP,           // in-model MTP/NextN head (Qwen3.5/3.6) loaded as a sibling ctx
```
Comment (am17an) [inline common/common.h]: simplify comment, to just `multi-token prediction`

File: common/speculative.cpp [hunk 3]
```diff
@@ -152,6 +154,13 @@ struct common_speculative_state {
 
     virtual void accept(uint16_t n_accepted) = 0;
 
+    // Optional hook: invoked by the server after each successful llama_decode
+    // on ctx_tgt. MTP uses it (only when is_prompt_prefill) to mirror the
+    // ubatch into ctx_mtp's KV.
+    virtual void on_target_decoded(const llama_batch & /*batch*/,
+                                   llama_seq_id /*slot_seq_id*/,
+                                   bool /*is_prompt_prefill*/) {}
+
```
Comment (am17an) [inline common/speculative.cpp]: this should be removed completely

File: common/speculative.cpp [hunk 4]
```diff
@@ -1199,11 +1438,29 @@ void common_speculative_accept(common_speculative * spec, uint16_t n_accepted) {
             impl->n_acc_tokens += n_accepted;
         }
 
+        // Always notify the impl, including for n_accepted=0. State_mtp needs
+        // this to update last_n_accepted (the row index used by the next k=0
+        // relay) and to trim ctx_mtp KV of all rejected drafts. Skipping the
+        // call leaves stale per-round bookkeeping which cascades into wrong
+        // h_pre_norm relays for subsequent draft rounds.
```
Comment (am17an) [inline common/speculative.cpp]: superflous comment

File: common/speculative.cpp [hunk 5]
```diff
@@ -1184,13 +1423,13 @@ llama_tokens common_speculative_draft(
...
-    }
-
     common_speculative_state * impl = spec->curr_impl;
 
-    GGML_ASSERT(impl);
+    if (!impl) {
+        // No draft was emitted last round (curr_impl is set inside draft()),
+        // so there's nothing to accept against. Safe to skip.
+        return;
+    }
```
Comment (am17an) [inline common/speculative.cpp:1367]: not required, n_accepted if block should be added back

File: common/speculative.cpp [hunk 6]
```diff
@@ -1199,11 +1438,29 @@ void common_speculative_accept(common_speculative * spec, uint16_t n_accepted) {
...
+        llama_seq_id         slot_seq_id,
+        bool                 is_prompt_prefill) {
+    if (!spec) {
+        return;
+    }
+    for (auto & impl : spec->impls) {
+        impl->on_target_decoded(batch, slot_seq_id, is_prompt_prefill);
+    }
+}
+
```
Comment (am17an) [inline common/speculative.cpp]: should not be required

File: src/models/models.h [hunk 7]
```diff
@@ -626,6 +626,17 @@ struct llm_build_qwen35 : public llm_build_delta_net_base {
     const llama_model & model;
 };
 
+// MTP draft head for Qwen3.5/3.6. A separate model loaded from the same .gguf
+// as the trunk via llama_model_params::override_arch = "qwen35_mtp". The graph
+// reads a hidden state staged on its input buffer (populated by an external
+// ggml_backend_tensor_copy_async from the target ctx's t_h_pre_norm) plus the
+// just-sampled token, projects through eh_proj, runs one full-attention block
+// (own KV cache via build_attn_inp_kv()), and emits draft logits via the shared
+// LM head.
```
Comment (am17an) [inline src/models/models.h]: remove slop comment

File: src/models/qwen35.cpp [hunk 8]
```diff
@@ -72,6 +74,15 @@ llm_build_qwen35::llm_build_qwen35(const llama_model & model, const llm_graph_pa
     }
     cur = inpL;
 
+    // Snapshot pre-output-norm hidden at output positions for MTP. A separate
+    // ctx_mtp (LLM_ARCH_QWEN35_MTP) pulls this tensor via
+    // ggml_backend_tensor_copy_async to feed its own MTP forward — h is the
+    // h_t the MTP head was trained against (DeepSeek V3 paper notation).
```
Comment (am17an) [inline src/models/qwen35.cpp]: remove comment

File: src/models/qwen35_mtp.cpp [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+    const int il = (int) hparams.n_layer - (int) hparams.nextn_predict_layers;
+    const auto & layer = model.layers[il];
+
+    GGML_ASSERT(layer.nextn.eh_proj && "MTP block missing nextn.eh_proj");
+    GGML_ASSERT(layer.nextn.enorm   && "MTP block missing nextn.enorm");
+    GGML_ASSERT(layer.nextn.hnorm   && "MTP block missing nextn.hnorm");
+
+    int sections[4];
+    std::copy(std::begin(hparams.rope_sections), std::begin(hparams.rope_sections) + 4, sections);
+
```
Comment (am17an) [inline src/models/qwen35_mtp.cpp:85]: is this standard?

File: src/llama-arch.h [hunk 10]
```diff
@@ -45,6 +45,7 @@ enum llm_arch {
     LLM_ARCH_QWEN3VLMOE,
     LLM_ARCH_QWEN35,
     LLM_ARCH_QWEN35MOE,
+    LLM_ARCH_QWEN35_MTP,    // MTP draft head for Qwen3.5/3.6 (loaded as a separate llama_model)
```
Comment (am17an) [inline src/llama-arch.h]: remove comment

File: src/llama-context.cpp [hunk 11]
```diff
@@ -3095,6 +3107,272 @@ void llama_set_warmup(llama_context * ctx, bool warmup) {
...
+
+        // Stage h-rows into ctx_mtp's t_inp_h source buffer.
+        const int32_t rc_relay = llama_mtp_relay_h(this, ctx_mtp,
+                                                   chunk_off, chunk_n);
+        GGML_ASSERT(rc_relay == 0 && "llama_mtp_relay_h failed");
+
+        // Build a single-seq batch over [chunk_pos_start..chunk_pos_start + chunk_n).
+        // Tokens are taken straight from the trunk ubatch (no-shift approximation:
+        // h_p paired with x_p, matching the existing on_target_decoded behavior).
+        llama_batch chunk = llama_batch_init(chunk_n, 0, 1);
```
Comment (am17an) [inline src/llama-context.cpp]: we definitely need to shift, what will be required here?

File: src/llama-context.cpp [hunk 12]
```diff
@@ -3095,6 +3107,272 @@ void llama_set_warmup(llama_context * ctx, bool warmup) {
...
+        return;
+    }
+
+    // Sub-chunk by ctx_mtp's n_ubatch.
+    const int n_ub_mtp = (int) llama_n_ubatch(ctx_mtp);
+    GGML_ASSERT(n_ub_mtp >= 1);
+
+    const int n_rows = (int) ubatch.n_tokens;
+    int chunk_off = 0;
+    while (chunk_off < n_rows) {
```
Comment (am17an) [inline src/llama-context.cpp]: can there be more than 1 chunk here? is that even possible? We can assert n_ubatch for mtp and target is the same

File: src/llama-graph.h [hunk 13]
```diff
@@ -120,6 +121,28 @@ class llm_graph_input_embd : public llm_graph_input_i {
...
 };
 
+// Graph input for the trunk's pre-output-norm hidden state. Populated during
+// set_input by reading the staged source on the owning llama_context (set by
+// llama_mtp_relay_h{,_self}) and doing a device-to-device copy from those
+// rows into this->h. Going through set_input rather than an immediate copy in
+// the relay function is what lets the relay survive a graph rebuild between
+// different n_tokens (e.g. n_tokens=N prompt prefill → n_tokens=1 chain step):
+// the staging lives on the context, but the destination this->h is whatever
+// tensor the current graph just allocated.
```
Comment (am17an) [inline src/llama-graph.h]: remove  comment

File: tools/server/server-context.cpp [hunk 14]
```diff
@@ -156,6 +164,16 @@ struct server_slot {
...
 
+    // Mirror a trunk-side seq_rm into ctx_mtp's KV (single-seq, so seq_id=0).
+    // No-op when MTP is not enabled for this slot. Keep this in lockstep with
+    // every llama_memory_seq_rm(ctx, id, ...) on the trunk side.
+    bool mtp_seq_rm(llama_pos p0, llama_pos p1) {
+        if (!ctx_mtp) {
+            return true;
+        }
+        return llama_memory_seq_rm(llama_get_memory(ctx_mtp), 0, p0, p1);
+    }
```
Comment (am17an) [inline tools/server/server-context.cpp]: why shouldn't this just be in inside `llama_memory_seq_rm`?

File: tools/server/server-context.cpp [hunk 15]
```diff
@@ -2636,12 +2754,16 @@ struct server_context_impl {
                             break;
                         }
 
-                        // embedding requires all tokens in the batch to be output
+                        // embedding requires all tokens in the batch to be output.
+                        // MTP draft also wants outputs at every prompt position
+                        // so common_speculative_begin can prefill ctx_mtp's KV
+                        // with t_h_pre_norm rows for the whole prompt.
+                        const bool is_mtp = params_base.speculative.type == COMMON_SPECULATIVE_TYPE_MTP;
```
Comment (am17an) [inline tools/server/server-context.cpp]: hmm maybe should be part of the slot? and need embd should return true?

File: src/llama-context.h [hunk 16]
```diff
@@ -69,6 +70,65 @@ struct llama_context {
...
+    void    clear_mtp();
+
+    // Last-row index (in the most recent ubatch's t_h_pre_norm) processed by
+    // the streaming hook. Used by state_mtp's first-draft k=0 relay so it can
+    // read the last-prompt-position row. -1 if not set.
+    int32_t get_mtp_last_h_row() const;
+
+    // Fired from inside process_ubatch after compute. Relays t_h_pre_norm rows
+    // into the registered ctx_mtp and decodes them into its KV.
+    void    handle_mtp_for_ubatch(const llama_ubatch & ubatch, ggml_tensor * t_h_pre_norm);
```
Comment (am17an) [inline src/llama-context.h]: this API surface is extremely large and doesn't seem to benefit much?

File: common/speculative.cpp (hunk 5 above)
Comment (am17an) [reply common/speculative.cpp:1367]: Not addressed yet — deferring. The fix is to restore the early-return and move the n=0 bookkeeping into `state_mtp` itself (detect lazily at the next `draft()` start by comparing `ctx_mtp`'s pos_max against expected). Keeps the public spec API clean.

File: common/speculative.cpp (hunk 6 above)
Comment (am17an) [reply common/speculative.cpp]: Still present after the refactor — will delete in a follow-up. Drop the C function, the virtual on the base class, and the header decl. The streaming hook in `process_ubatch` covers everything this used to do.

File: src/models/models.h (hunk 7 above)
Comment (am17an) [reply src/models/models.h]: Done in e821ca066.

File: src/models/qwen35.cpp (hunk 8 above)
Comment (am17an) [reply src/models/qwen35.cpp]: Done in e821ca066.

File: src/models/qwen35_mtp.cpp (hunk 9 above)
Comment (am17an) [reply src/models/qwen35_mtp.cpp:85]: Yes — copied verbatim from `qwen35.cpp` / `qwen35moe.cpp`. Same pattern in every M-RoPE builder (qwen2vl, qwen3vl{,moe}, glm4{,moe}, hunyuan-dense, paddleocr).

File: src/llama-arch.h (hunk 10 above)
Comment (am17an) [reply src/llama-arch.h]: Done in e821ca066.

File: src/llama-context.cpp (hunk 11 above)
Comment (am17an) [reply src/llama-context.cpp]: Not addressed yet — `handle_mtp_for_ubatch` still pairs `(h_p, x_p)` after the 0a837d2ef refactor. Plan for the shift fix: skip pos 0 (no `h_{-1}`); maintain a single `mtp.pending_h` row across ubatches so the boundary token from the next ubatch can be paired with the current ubatch's last h; the trailing pending h after final prefill is consumed naturally by `state_mtp::draft`'s k=0 step.

File: src/llama-context.cpp (hunk 12 above)
Comment (am17an) [reply src/llama-context.cpp]: Done in 0a837d2ef. `set_mtp` now asserts `n_ubatch(ctx_mtp) == n_ubatch(trunk)` and `handle_mtp_for_ubatch` decodes the whole ubatch as a single batch — no chunking loop.

File: src/llama-graph.h (hunk 13 above)
Comment (am17an) [reply src/llama-graph.h]: Done in e821ca066. Class itself was deleted entirely in 0a837d2ef.

File: tools/server/server-context.cpp (hunk 14 above)
Comment (am17an) [reply tools/server/server-context.cpp]: Not addressed yet. Will move the mirror into `llama_context::memory_seq_rm` so the trunk's `seq_rm` auto-dispatches to the registered ctx_mtp; eliminates the foot-gun of the server forgetting to mirror at any of the call sites.

File: tools/server/server-context.cpp (hunk 15 above)
Comment (am17an) [reply tools/server/server-context.cpp]: Not addressed yet. Will fold `is_mtp` into `slot.task->need_embd()` so the OR at this call site disappears and any other `need_embd` check gets the right answer for free.

File: src/llama-context.h (hunk 16 above)
Comment (am17an) [reply src/llama-context.h]: Mostly done in 0a837d2ef. Removed: `mtp_h_source_t`, `set/get/clear_mtp_h_source`, `get_t_inp_h`, `llama_mtp_relay_h`, `llama_mtp_relay_h_self`, the `llama_context * ctx` field on `llm_graph_params`, the `t_inp_h` field on `llm_graph_result`, the custom `llm_graph_input_h_pre_norm` class. h now flows through `batch.embd` → `llm_graph_input_embd::embd` — the standard input path, no MTP-specific staging surface. Net −281 LoC across 8 files. Public surface remaining: `llama_set_mtp` / `llama_clear_mtp` / `llama_get_mtp_last_h_row` / `llama_context_get_t_h_pre_norm` / `llama_context_get_t_mtp_out`. Verified bit-identical against pre-refactor on an 8-prompt fixed-seed bench (every per-prompt draft/accept count matches exactly).

File: common/speculative.cpp (hunk 5 above)
Comment (am17an) [reply common/speculative.cpp:1367]: Done in c92a70c77. Restored `if (n_accepted == 0) return;`. Moved the n=0 bookkeeping into `state_mtp::draft`: at the start of each draft round it detects a skipped `accept(0)` via leftover `last_n_drafted > 0`, trims the K-1 stale draft positions in ctx_mtp's KV, and advances `last_n_accepted` to 0 before the relay.

File: common/speculative.cpp (hunk 6 above)
Comment (am17an) [reply common/speculative.cpp]: Done in c92a70c77. Deleted the C function, the virtual on the base class, and the header decl.

File: src/llama-context.cpp (hunk 11 above)
Comment (am17an) [reply src/llama-context.cpp]: Done in c92a70c77. `handle_mtp_for_ubatch` now does `(h_p, x_{p+1})` pairing: skips MTP pos 0 (no `h_{-1}`), tracks a single pending h-row across ubatches via `mtp.pending_h`/`pending_pos`, consumes pending in next ubatch's first slot, and stashes new pending after the call. The trailing pending after final prefill is naturally consumed by `state_mtp::draft`'s k=0 step. Bench: aggregate accept 0.7996 → 0.825 (+2.5%); the multi-ubatch `long_code_review` prompt 0.688 → 0.745 (+5.8%) — the shift's intended target.

File: tools/server/server-context.cpp (hunk 14 above)
Comment (am17an) [reply tools/server/server-context.cpp]: Done in c92a70c77. New `llama_context_seq_rm(ctx, seq_id, p0, p1)` wraps `llama_memory_seq_rm` and auto-mirrors to ctx_mtp (single-seq, id 0). Server replaced 7 trunk `llama_memory_seq_rm` call sites with `llama_context_seq_rm`; the explicit `slot.mtp_seq_rm()` helper and its mirror calls (prompt_clear, copy_state_to, ctx_shift, n_cache_reuse, truncate, checkpoint-restore, accepted-tokens-trim) are gone.

File: tools/server/server-context.cpp (hunk 15 above)
Comment (am17an) [reply tools/server/server-context.cpp]: Done in c92a70c77. Added `slot.is_mtp()` (returns `ctx_mtp != nullptr`) and `slot.need_embd()` which folds in the OR. Both call sites (batch fill at L2761, `llama_set_embeddings` at L2878) now use `slot.need_embd()`; the local `is_mtp` variable is gone.
